### DISCUSSION TITLE
Requests fail if coming from within another container

### DISCUSSION THREAD

**Discussion by Hi-ImKyle** [NginxProxyManager/nginx-proxy-manager#3966] [Q&A]
So I've had this issue for a while now and I've only recently needed it fixing so I'm asking here since I think it's semi related to Nginx/PM the most.

To sum it up, I have NPM in a docker container. Everything works. All my proxies work fine from the outside. So if I go to `xyz.my-domain.com` it brings up whatever it is at `xyz`. But if I curl that url from within another container, it resolves my domain to my IP address then says "Trying my-ip:443" and eventually just times out.

This unfortunately means I can't setup Authentik's OpenID Connect in services that support it since I can't access my domain from inside a container. And before you ask, yes I can access authentik from within another container if its on the same network but that causes authentik to provide the wrong urls.

This issue isn't just with Authentik. It doesn't matter what a container is hosting, it doesn't work using it's domain when curl'ing said domain from another container.

Is this a Docker issue, Nginx/PM issue or what? I just don't understand what could cause it to time out like this.

For those curious, here is `curl -vk https://auth.my-domain.com/` from within Gitea.
```
7a229214a1a5:/# curl -vk https://auth.my-domain.com/
* Host auth.my-domain.com:443 was resolved.
* IPv6: (none)
* IPv4: 1.2.3.4
*   Trying 1.2.3.4:443...
* connect to 1.2.3.4 port 443 from 192.168.32.3 port 52394 failed: Operation timed out
* Failed to connect to auth.my-domain.com port 443 after 130328 ms: Couldn't connect to server
* Closing connection
curl: (28) Failed to connect to auth.my-domain.com port 443 after 130328 ms: Couldn't connect to server
```

**Comment by JonhyOliveira** [comment]
Have you tried looking at NPM access logs?

**Reply by Hi-ImKyle** [reply]
Yes. No logs related to any connection failing/timing out or any connection from any other IP than my desktop.

**Reply by JonhyOliveira** [reply]
If Authentik's requests don't show up in any NPM access log you need to look elsewhere. Can you curl from within the service to itself (curl xyz.example.com)?

Seems to me like some weird routing is going on

**Reply by Hi-ImKyle** [reply]
It's not Authentik that is the problem. Authentik isn't the one requesting anything. It's just an example. 

Like I can't tell my Gitea `git.example.com` to use the OAuth that Authentik offers at `auth.example.com`, because the Gitea container just always says there was a timeout whenever trying :443. But you can substitute these contains for anything. It doesn't matter what they are hosting, getting a container to request anything from another container using the domain tied to the other container **always** times out.

Let's say I have Container A and Container B. Container B is hosting images at `img.example.com`. And from within Container A, I try to curl, fetch, wget, whatever HTTP requesting system. I try to get `img.example.com/xyz.png` it will resolve the domain correctly to the IP address. Then try connecting to said IP at :443, but times out.

I only say I think it's nginx/pm because that's the container on 443 that is routing things.

**Reply by JonhyOliveira** [reply]
I don't think Authentik is the problem either but I'm not sure if NPM is. 

We should first figure out if there are any connectivity issues between container A and the NPM container, though.

From this :point_down:  it makes me think that there is a connectivity issue, because the connection is timing out. If it's not showing up on any logs it means that it's not received by NPM and is timing out elsewhere (your packets are not being delivered)
> Yes. No logs related to any connection failing/timing out or any connection from any other IP than my desktop.

I'm running the same setup and have no issues reaching the host on container A from the host on container B.

Show the `ping` results of pinging (via IP), from within one of the docker containers:
1. Your router
2. Your host
3. The NPM container

If all pings are sucessful you should at least see logs of your cURL requests on NPM.

**Reply by JonhyOliveira** [reply]
Run your ping program 2-3 times for each one. If all pings are successful then try pinging the other container via IP.

**Reply by Hi-ImKyle** [reply]
I don't think you fully understand my problem but I have done what you asked anyway.
Container A pinging the router:
```
7a229214a1a5:/# ping -c 4 192.168.32.1
PING 192.168.32.1 (192.168.32.1): 56 data bytes
64 bytes from 192.168.32.1: seq=0 ttl=64 time=0.066 ms
64 bytes from 192.168.32.1: seq=1 ttl=64 time=0.057 ms
64 bytes from 192.168.32.1: seq=2 ttl=64 time=0.056 ms
64 bytes from 192.168.32.1: seq=3 ttl=64 time=0.057 ms

--- 192.168.32.1 ping statistics ---
4 packets transmitted, 4 packets received, 0% packet loss
round-trip min/avg/max = 0.056/0.059/0.066 ms
```
Container A pinging NPM
```
7a229214a1a5:/# ping -c 4 nginx-app-1
PING nginx-app-1 (192.168.32.6): 56 data bytes
64 bytes from 192.168.32.6: seq=0 ttl=64 time=0.118 ms
64 bytes from 192.168.32.6: seq=1 ttl=64 time=0.098 ms
64 bytes from 192.168.32.6: seq=2 ttl=64 time=0.096 ms
64 bytes from 192.168.32.6: seq=3 ttl=64 time=0.086 ms

--- nginx-app-1 ping statistics ---
4 packets transmitted, 4 packets received, 0% packet loss
round-trip min/avg/max = 0.086/0.099/0.118 ms
```
Container A pinging the host:
```
7a229214a1a5:/# ping -c 4 1.2.3.4
PING 1.2.3.4 (1.2.3.4): 56 data bytes
64 bytes from 1.2.3.4: seq=0 ttl=64 time=0.056 ms
64 bytes from 1.2.3.4: seq=1 ttl=64 time=0.049 ms
64 bytes from 1.2.3.4: seq=2 ttl=64 time=0.054 ms
64 bytes from 1.2.3.4: seq=3 ttl=64 time=0.052 ms

--- 1.2.3.4 ping statistics ---
4 packets transmitted, 4 packets received, 0% packet loss
round-trip min/avg/max = 0.049/0.052/0.056 ms
```

Pings are fine.

**Reply by JonhyOliveira** [reply]
> I don't think you fully understand my problem but I have done what you asked anyway. Container A pinging the router:

That is the case, but isn't that what we're doing here? Trying to find out why you're getting this timeout.

> 
> ```
> 7a229214a1a5:/# ping -c 4 192.168.32.1
> PING 192.168.32.1 (192.168.32.1): 56 data bytes
> 64 bytes from 192.168.32.1: seq=0 ttl=64 time=0.066 ms
> 64 bytes from 192.168.32.1: seq=1 ttl=64 time=0.057 ms
> 64 bytes from 192.168.32.1: seq=2 ttl=64 time=0.056 ms
> 64 bytes from 192.168.32.1: seq=3 ttl=64 time=0.057 ms
> 
> --- 192.168.32.1 ping statistics ---
> 4 packets transmitted, 4 packets received, 0% packet loss
> round-trip min/avg/max = 0.056/0.059/0.066 ms
> ```
> 
> Container A pinging NPM
> 
> ```
> 7a229214a1a5:/# ping -c 4 nginx-app-1
> PING nginx-app-1 (192.168.32.6): 56 data bytes
> 64 bytes from 192.168.32.6: seq=0 ttl=64 time=0.118 ms
> 64 bytes from 192.168.32.6: seq=1 ttl=64 time=0.098 ms
> 64 bytes from 192.168.32.6: seq=2 ttl=64 time=0.096 ms
> 64 bytes from 192.168.32.6: seq=3 ttl=64 time=0.086 ms
> 
> --- nginx-app-1 ping statistics ---
> 4 packets transmitted, 4 packets received, 0% packet loss
> round-trip min/avg/max = 0.086/0.099/0.118 ms
> ```

Something wrong may be happening here. This - `PING nginx-app-1 (192.168.32.6)` - indicates that container A is resolving nginx-app-1 to your LAN IP. It should be resolving to an IP in the docker network that is shared with nginx (by default it'll be a net in the 172.16.0.0/12 range). 

Here's what I see on my machine, where I don't have the issue you do:

```
4bcb674db1c3:/config# ping nginx-app-1
PING nginx-app-1 (172.18.0.4): 56 data bytes
64 bytes from 172.18.0.4: seq=0 ttl=64 time=0.077 ms
64 bytes from 172.18.0.4: seq=1 ttl=64 time=0.058 ms
64 bytes from 172.18.0.4: seq=2 ttl=64 time=0.075 ms
```

As you can see the name gets correctly resolved an IP on the dedicated docker network.

Sharing a `docker network inspect` of the network would probably help debug this further.

Also, try pinging container B from container A and see if the IP is still resolved to the host machine's LAN IP.

**Reply by JonhyOliveira** [reply]
> I can access authentik from within another container if its on the same network but that causes authentik to provide the wrong urls.

I was re-reading your original post and this is probably expected (not a docker expert). I also don't really understand what these "wrong urls" are. You should probably look into fixing the process that generates those URLs. Doesn't seem to me like having no urls is a step in the right direction.

**Reply by Hi-ImKyle** [reply]
Authentik generates some OpenID authentication endpoint urls based on the request to the auto discovery URL. Depending on the domain used in this request, the output of the endpoint urls is changed to show the domain from the request and not from any setting.

If I request these endpoint urls using `http://authentik-server-1:9000/...` it gives all the endpoint urls the `http://authentik-server-1:9000` domain, which is bad because some apps use these urls to redirect the user with. And they won't work this way. But if I got them using `https://auth.example.com/...` they would have the `auth.example.com` domain which is correct for accessing Authentik from outside of the host.

This to me doesn't really change the problem I have. Because eventually I will still need to access things in the way I'm saying doesn't work in the future, changing Authentik to hard code domains to fix this authentication issue is just a bandaid fix for something else.

All containers can ping each other fine and can resolve my domain fine. All containers time out when trying to connect to public-ip:443.

If the request isn't even getting to nginx/pm, I'm just going to assume it's a docker issue somewhere, some config I've missed. IDK.

Thanks for trying anyway. If you have any further suggestions, I'm all ears..

**Reply by JonhyOliveira** [reply]
> Authentik generates some OpenID authentication endpoint urls based on the request to the auto discovery URL. Depending on the domain used in this request, the output of the endpoint urls is changed to show the domain from the request and not from any setting.
> 
> If I request these endpoint urls using `http://authentik-server-1:9000/...` it gives all the endpoint urls the `http://authentik-server-1:9000` domain, which is bad because some apps use these urls to redirect the user with. And they won't work this way. But if I got them using `https://auth.example.com/...` they would have the `auth.example.com` domain which is correct for accessing Authentik from outside of the host.

I think it's right to conclude that you MUST make a request to `https://auth.example.com` since, as you describe, Authentik is seemingly using the "Host" HTTP header to generate URLs (as it should). In case you're not aware, making an HTTP request to `example.com` results in a request to `example.com`'s IP + setting the "Host" header to `example.com`.

What's weird to me is why you can't, with the setup of having all containers on the same net, make a request to `https://auth.example.com` and get a `auth.example.com` URL. 

What comes to mind is the fact that nginx _can_ mangle the "Host" header so that may be what is happening here. The issue your describing actually lines up with this, since the common configuration for reverse proxies is to have a `X-Forwarded-Host` replace `Host` (auth.example.com) and `Host` gets set to the target hostname (authentik-server-1).

You might have to find a way to configure Authentik to let it know that it's behind a proxy server. Here's a Authentik page describing this: https://docs.goauthentik.io/docs/installation/reverse-proxy
Haven't looked trough it myself, but should be a good place to start.

**Reply by Hi-ImKyle** [reply]
Again, you are not understanding my problem. I can't make a request to `https://auth.example.com`, it doesn't get through, it always times out.

I can make that request locally on my desktop, which is a completely different machine. That works fine. But if I make the request on ANY container, it ALWAYS times out. Nginx doesn't log anything about it timing out and Authentik never receives the request in the first place.

To make the point clear. This is NOT an Authentik issue. I am using Authentik as an example of a service that containers cannot access using the domain provided by Nginx/PM.

**Reply by JonhyOliveira** [reply]
For me connectivity is a much bigger problem than wrongly formatted URLs. In your first post you describe a setup where you had the connectivity problem solved so I don't understand why you're so adamant in staying away from that setup.

If you don't acknowledge the problem I've identified as a problem then I can only wish you good luck in finding the real one.

Cheers

**Reply by Hi-ImKyle** [reply]
You are correct, connectivity is a bigger problem. Which is the main problem.

The solution I originally had doesn't work for everything. In fact it doesn't really work at all. I'm not going to use that bandaid fix for everything just because it lets containers connect fine. Like cool, that fixed that issue but now theres another issue of containers redirecting users to the completely wrong url and asking people to go "oh right yeah, just change that url there to xyz **every time you login**" is a hassle and stupid.

I'm sorry you feel the way you do. I'm still not convinced you understand my issue at all.

**Reply by JonhyOliveira** [reply]
I hope you got something out of giving me attitude but it's definitely not a productive response..

I'll still address the main points you left here anyway for your own sake and in case you want to actually get help.

> The solution I originally had doesn't work for everything. In fact it doesn't really work at all.

Just because you got it to work using the configuration you have now doesn't mean it's adequate.

How can you hope to restore connectivity if you are unwilling to touch your network configuration? If you have no connectivity this is also not an NPM issue.

I've explained this to you in this response: https://github.com/NginxProxyManager/nginx-proxy-manager/discussions/3966#discussioncomment-10514916; and the one after.

> I'm not going to use that bandaid fix for everything just because it lets containers connect fine.

Having all containers in the same network is NOT a bandaid fix! It's the ["Docker approved"](https://forums.docker.com/t/how-to-create-a-network-of-containers-that-can-communicate-with-each-other-interchangably/134292) way of configuring docker to have containers communicate with each other.

> Like cool, that fixed that issue but now theres another issue of containers redirecting users to the completely wrong url and asking people to go "oh right yeah, just change that url there to xyz **every time you login**" is a hassle and stupid.

That's not at all what I've been suggesting... I'm suggesting that AFTER fixing your network issue you can configure nginx to fix the URL issue.

Some feedback: Your lack of experience is clear so you shouldn't be shunning away people who are trying to assist you just because the solution they give you does not fit with your pre-concived notion of what the solution should be. Try to be more open to suggestions next time you ask for help.

**Reply by Hi-ImKyle** [reply]
If you feel that anything I said was me giving you attitude, you are looking into it wrongly. But that's ok.

You have tried to help me and it didn't work. Thanks.
